lib/squadster/formations/squad_member.ex

```diff
@@ -23,16 +26,53 @@ defmodule Squadster.Formations.SquadMember do
     struct
     |> cast(params, [:role, :queue_number, :user_id, :squad_id])
     |> validate_required([:role, :user_id, :squad_id])
-    |> reset_queue_number_if_commander
+    |> reset_queue_number
     |> foreign_key_constraint(:user_id)
     |> foreign_key_constraint(:squad_id)
   end
 
-  def reset_queue_number_if_commander(%Ecto.Changeset{changes: %{role: :commander}} = changeset) do
-    changeset |> put_change(:queue_number, nil)
+  defp reset_queue_number(%Ecto.Changeset{data: member, changes: %{role: new_role}} = changeset) do
+    cond do
+      new_role |> should_be_on_duty? && member |> in_queue? -> changeset
+      new_role |> should_be_on_duty? -> changeset |> to_end
+      true -> changeset |> remove_from_queue
+    end
+  end
+
+  defp reset_queue_number(changeset), do: changeset
+
+  defp should_be_on_duty?(role) do
+    cond do
+      role in [:commander, :deputy_commander, :journalist] -> false
+      true -> true
+    end
```

Comment:
What does it mean??
Why do you do this in that way

Reply:
this means else return true